torch/csrc/jit/codegen/cuda/compute_at_map.cpp

```diff
@@ -249,8 +262,8 @@ findFirstSelfMapping(Fusion* fusion, const IterDomainGraph& id_graph) {
     }
 
     // Leaf domains
-    auto self_mappped_leaf_pair =
-        detectMappablePair(tv->domain()->domain(), id_graph);
+    auto self_mappped_leaf_pair = detectMappablePair(
+        tv->domain()->domain(), id_graph, IdMappingMode::LOOP);
```

Comment:
I see that why the PERMISSIVE mapping is too strict for root domains.

For leaf domains, I also see that CA domains should not be even permissively mapped. But LOOP doesn't map anything for domains right of CA points. I can't think of anything that could actually trigger self mapping in those domains, though, as at this point we already have verified there's no self mapping in root domains, so I *think* checking with LOOP is sufficient.

Let me know if I misunderstood anything.

Reply:
So you think we should just check leaf nodes here, nothing else? I thought the self mapping logic got moved primarily to compute at map, so I thought we need a check on root/rfactor too.

Reply:
No, we should check root domains. I was wondering why we use LOOP for leaves instead of EXACT. For CA leaf domains, even permissive mappings shouldn't be allowed, which LOOP verifies. For non-CA leaf domains (right of CA point), LOOP won't catch anything as it doesn't map anything, effectively meaning we don't do any check of non-CA leaf domains. I think that's fine as long as roots are checked.

Reply:
To be clear, the PR makes sense to me. I just want to confirm I understand it correctly.

Reply:
> I was wondering why we use LOOP for leaves instead of EXACT.

I think it makes sense to use LOOP instead of EXACT. As long as different for loops in the same loop nest are not mapped, we don't care if they are mapped or not. For example, for the following fusion

```
T0[I1, I2] = input
T1 = transpose(T0, 0, 1) 
T2 = T1 + T0
output = T2
```

I think the exact map has self-mapping, but as long as we are not inlining, that will not be a problem: We can generate code
```
float T1[I1, I2];
for i2 in I2:
  for i1 in I1:
    T1[i2, i1] = T1[i1, i2]
for i2 in I2:
  for i1 in I1:
    T1[i2, i1] = T1[i2, i1]
```
And this code is totally valid

(probably we should add the above example as a unit test?)

Reply:
Yeah, this is more about historical technical debt that it has been generally assumed that there's no mapping between domains in a tensor domain. Although I don't know if the assumption is still there, just lifting it without making sure there's no code based on that assumption seems risky.

Reply:
Just checked the following test with this branch

```C++
TEST_F(NVFuserTest, FusionSelfMappingExactMap_CUDA) {
  // Check that even if there is self mapping in the exact compute at map, as
  // long as there is no self mapping in the loop map, we will be fine.
  auto fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  auto tv0 = makeConcreteTensor({5, 5});
  fusion.addInput(tv0);

  auto tv1 = transpose(tv0, 0, 1);
  auto tv2 = add(tv0, tv1);

  fusion.addOutput(tv2);

  // no inlining

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  auto t0 = at::randn({5, 5}, options);

  FusionExecutor fe;
  fe.compileFusion(&fusion, {t0});
  auto cg_outputs = fe.runFusion({t0});

  auto ref = t0.t() + t0;

  testValidate(&fusion, cg_outputs, {t0}, {ref}, __LINE__, __FILE__);
}
```
I hit a `TORCH_INTERNAL_ASSERT` in indexing. After removing all failing `TORCH_INTERNAL_ASSERT`s, I start to get segfault. So yes, we do have lot of technical debts that prevent us from supporting this case.